File: src/main/java/me/zeroeightsix/kami/gui/UIRenderer.java
package me.zeroeightsix.kami.gui;

import me.zeroeightsix.kami.KamiMod;
import me.zeroeightsix.kami.gui.kami.DisplayGuiScreen;
import me.zeroeightsix.kami.gui.kami.KamiGUI;
import me.zeroeightsix.kami.gui.rgui.component.Component;
import me.zeroeightsix.kami.gui.rgui.component.container.use.Frame;
import me.zeroeightsix.kami.gui.rgui.component.listen.RenderListener;
import me.zeroeightsix.kami.util.Wrapper;
import net.minecraft.client.renderer.GlStateManager;
import org.lwjgl.opengl.GL11;

public class UIRenderer {

    public static void renderAndUpdateFrames() {
        if (Wrapper.getMinecraft().currentScreen instanceof DisplayGuiScreen || Wrapper.getMinecraft().gameSettings.showDebugInfo)
            return;
        KamiGUI gui = KamiMod.getInstance().getGuiManager();

        GL11.glDisable(GL11.GL_TEXTURE_2D);
        for (Component c : gui.getChildren()) {
            if (c instanceof Frame) {
                GlStateManager.pushMatrix();
                Frame child = (Frame) c;
                if (child.isPinned() && child.isVisible()) {
                    boolean slide = child.getOpacity() != 0;
                    GL11.glTranslated(child.getX(), child.getY(), 0);
                    child.getRenderListeners().forEach(renderListener -> renderListener.onPreRender());
                    child.getTheme().getUIForComponent(child).renderComponent(child, child.getTheme().getFontRenderer());

                    int translateX = 0;
                    int translateY = 0;

                    if (slide) {
                        translateX += child.getOriginOffsetX();
                        translateY += child.getOriginOffsetY();
                    } else {
                        if (child.getDocking().isBottom()) {
                            translateY += child.getOriginOffsetY();
                        }
                        if (child.getDocking().isRight()) {
                            translateX += child.getOriginOffsetX();
                            if (child.getChildren().size() > 0) {
                                translateX += (child.getWidth() - child.getChildren().get(0).getX() - child.getChildren().get(0).getWidth()) / DisplayGuiScreen.getScale();
                            }
                        }
                        if (child.getDocking().isLeft() && child.getChildren().size() > 0) {
                            translateX -= child.getChildren().get(0).getX();
                        }
                        if (child.getDocking().isTop() && child.getChildren().size() > 0) {
                            translateY -= child.getChildren().get(0).getY();
                        }
                    }

                    GL11.glTranslated(translateX, translateY, 0);
                    child.getRenderListeners().forEach(RenderListener::onPostRender);
                    child.renderChildren();
                    GL11.glTranslated(-translateX, -translateY, 0);
                    GL11.glTranslated(-child.getX(), -child.getY(), 0);
                }
                GlStateManager.popMatrix();
            }
        }
        GL11.glEnable(GL11.GL_TEXTURE_2D);
        GL11.glEnable(GL11.GL_BLEND);
        GlStateManager.enableBlend();
    }

}

File: src/main/java/me/zeroeightsix/kami/gui/UIRenderer.kt
package me.zeroeightsix.kami.gui

import me.zeroeightsix.kami.KamiMod
import me.zeroeightsix.kami.gui.kami.DisplayGuiScreen
import me.zeroeightsix.kami.gui.rgui.component.container.use.Frame
import me.zeroeightsix.kami.util.Wrapper
import me.zeroeightsix.kami.util.graphics.GlStateUtils
import org.lwjgl.opengl.GL11.*

object UIRenderer {
    fun renderAndUpdateFrames() {
        if (Wrapper.minecraft.currentScreen is DisplayGuiScreen || Wrapper.minecraft.gameSettings.showDebugInfo) return
        val gui = KamiMod.INSTANCE.guiManager
        GlStateUtils.rescaleKami()

        for (child in gui.children) {
            if (child !is Frame) continue
            if (!child.isPinned || !child.isVisible) continue

            glPushMatrix()
            val slide = child.opacity != 0f
            glTranslated(child.x.toDouble(), child.y.toDouble(), 0.0)
            for (renderListener in child.renderListeners) renderListener.onPreRender()
            child.theme.getUIForComponent(child).renderComponent(child)
            var translateX = 0.0
            var translateY = 0.0
            if (slide) {
                translateX += child.originOffsetX
                translateY += child.originOffsetY
            } else {
                if (child.docking.isBottom) {
                    translateY += child.originOffsetY
                }
                if (child.docking.isRight) {
                    translateX += child.originOffsetX
                    if (child.children.size > 0) {
                        translateX += (child.width - child.children[0].x - child.children[0].width) / DisplayGuiScreen.getScale()
                    }
                }
                if (child.docking.isLeft && child.children.size > 0) {
                    translateX -= child.children[0].x
                }
                if (child.docking.isTop && child.children.size > 0) {
                    translateY -= child.children[0].y
                }
            }
            glTranslated(translateX, translateY, 0.0)
            for (renderListener in child.renderListeners) renderListener.onPostRender()
            child.renderChildren()
            glPopMatrix()
        }
        GlStateUtils.rescaleMc()
    }
}